File: src/aos.cpp
#include "aos.h"

namespace aos {

ParticleSystem::ParticleSystem(size_t count) {
    particles.reserve(count);
    for (size_t i = 0; i < count; ++i) {
        float fi = static_cast<float>(i);
        particles.emplace_back(
            Vec3(fi, fi * 2.0f, fi * 3.0f),           // position
            Vec3(fi * 0.1f, fi * 0.2f, fi * 0.3f),    // velocity
            1.0f + fi * 0.01f                          // mass
        );
    }
}

void ParticleSystem::update_positions(float dt) {
    for (auto& particle : particles) {
        particle.position = particle.position.add(particle.velocity.scale(dt));
    }
}

float ParticleSystem::compute_kinetic_energy() const {
    float total = 0.0f;
    for (const auto& particle : particles) {
        const Vec3& v = particle.velocity;
        float v2 = v.x * v.x + v.y * v.y + v.z * v.z;
        total += 0.5f * particle.mass * v2;
    }
    return total;
}

void ParticleSystem::apply_gravity(const Vec3& gravity, float dt) {
    for (auto& particle : particles) {
        particle.velocity = particle.velocity.add(gravity.scale(dt));
    }
}

float ParticleSystem::update(const Vec3& gravity, float dt) {
    apply_gravity(gravity, dt);
    update_positions(dt);
    return compute_kinetic_energy();
}

} // namespace aos

File: src/aos.rs
/// Array of Structures (AoS) - Cache Unfriendly
/// When we iterate to update positions, we skip over velocity and mass data,
/// leading to poor cache utilization

use crate::Vec3;

#[derive(Clone, Debug)]
pub struct Particle {
    pub position: Vec3,
    pub velocity: Vec3,
    pub mass: f32,
}

impl Particle {
    pub fn new(position: Vec3, velocity: Vec3, mass: f32) -> Self {
        Self {
            position,
            velocity,
            mass,
        }
    }
}

pub struct ParticleSystem {
    pub particles: Vec<Particle>,
}

impl ParticleSystem {
    pub fn new(count: usize) -> Self {
        let mut particles = Vec::with_capacity(count);
        for i in 0..count {
            let fi = i as f32;
            particles.push(Particle::new(
                Vec3::new(fi, fi * 2.0, fi * 3.0),
                Vec3::new(fi * 0.1, fi * 0.2, fi * 0.3),
                1.0 + fi * 0.01,
            ));
        }
        Self { particles }
    }

    /// Update particle positions based on velocity
    /// Poor cache behavior: we load entire Particle struct (40 bytes) but only need
    /// position (12 bytes) and velocity (12 bytes)
    #[inline(never)]
    pub fn update_positions(&mut self, dt: f32) {
        for particle in &mut self.particles {
            particle.position = particle.position.add(&particle.velocity.scale(dt));
        }
    }

    /// Compute total kinetic energy
    /// Poor cache behavior: we access velocity and mass, skipping position data
    #[inline(never)]
    pub fn compute_kinetic_energy(&self) -> f32 {
        let mut total = 0.0;
        for particle in &self.particles {
            let v2 = particle.velocity.x * particle.velocity.x
                + particle.velocity.y * particle.velocity.y
                + particle.velocity.z * particle.velocity.z;
            total += 0.5 * particle.mass * v2;
        }
        total
    }

    /// Apply gravity to all particles
    /// Poor cache behavior: we only need to modify velocity, but load entire struct
    #[inline(never)]
    pub fn apply_gravity(&mut self, gravity: Vec3, dt: f32) {
        for particle in &mut self.particles {
            particle.velocity = particle.velocity.add(&gravity.scale(dt));
        }
    }
}
